test/evp_pkey_provided_test.c

```diff
@@ -443,6 +443,362 @@ static int test_fromdata_rsa(void)
     return ret;
 }
 
+struct check_data {
+    const char *pname;
+    BIGNUM *comparebn;
+};
+
+
+static int do_fromdata_rsa_derive(OSSL_PARAM *fromdata_params, struct check_data check[],
+                                    int expected_nbits, int expected_sbits, int expected_ksize)
+{
+    const OSSL_PARAM *check_param = NULL;
+    BIGNUM *check_bn = NULL;
+    OSSL_PARAM *todata_params = NULL;
+    EVP_PKEY_CTX *ctx = NULL, *key_ctx = NULL;
+    EVP_PKEY *pk = NULL, *copy_pk = NULL, *dup_pk = NULL;
+    int i;
+    int ret = 0;
+
+    if(!TEST_ptr(ctx = EVP_PKEY_CTX_new_from_name(NULL, "RSA", NULL))
+        || !TEST_int_eq(EVP_PKEY_fromdata_init(ctx), 1)
+        || !TEST_int_eq(EVP_PKEY_fromdata(ctx, &pk, EVP_PKEY_KEYPAIR,
+                                          fromdata_params), 1))
+        goto err;
+
+    /*
+     * get the generated key parameters back and validate that the
+     * exponents/coeffs are correct
+     */
+    if (!TEST_int_eq(EVP_PKEY_todata(pk, EVP_PKEY_KEYPAIR, &todata_params), 1))
+        goto err;
+
+    for (i=0; check[i].pname; i++) {
+        if (!TEST_ptr(check_param = OSSL_PARAM_locate_const(todata_params, check[i].pname)))
+            goto err;
+        if (!TEST_int_eq(OSSL_PARAM_get_BN(check_param, &check_bn), 1))
+            goto err;
+        if (!TEST_BN_eq(check_bn, check[i].comparebn)) {
+            TEST_info("Data mismatch for parameter %s\n", check[i].pname);
+            goto err;
+        }
+        BN_free(check_bn);
+        check_bn = NULL;
+    }
+
+    while (dup_pk == NULL) {
```

Comment:
Can you explain the purpose of this loop?

Reply:
I think this was the question you didnt answer..

Reply:
dup_pk is non NULL at the end of this block so it only runs once? Was this the intention

Reply:
Oh yeah, this pattern was copied from test_fromdata_rsa and is wrong (and elsewhere too). It needs to be a `for (;;)` loop which breaks out in the middle before calling `EVP_PKEY_dup()` when `dup_pk != NULL`.

Reply:
This one doesnt seem to be addressed?